Spec edits for incremental delivery, Response Section and Examples Appendix (#1203)
graphql/graphql-spec: spec/Appendix E -- Examples.md

responses.

```json example
{
  "data": {
    "person": {
      "firstName": "Luke"
    }
  },
  "pending": [
    { "id": "0", "path": ["person"], "label": "homeWorldDefer" },
    { "id": "1", "path": ["person"], "label": "nameAndWorld" }
  ],
  "hasNext": true
}
```

Subsequent response 1, contains the deferred data from `HomeWorldFragment`.
There is one Completed Result, indicating that `HomeWorldFragment` has been
completely delivered. Because the `homeWorld` field is present in two separate
`@defer`s, it is separated into its own Incremental Result.

The second Incremental Result contains the data for the `terrain` field. This
incremental result contains a `subPath` property to indicate to clients that the
path of this result can be determined by concatenating the path from the Pending
Result with id `"0"` and this `subPath` entry.

```json example
{
  "incremental": [
    {
      "id": "0",
      "data": { "homeWorld": { "name": "Tatooine" } }
    },
    {
      "id": "0",
      "subPath": ["homeWorld"],
      "data": { "terrain": "desert" }
    }
  ],
  "completed": [{ "id": "0" }],
  "hasNext": true
}
```

> Subsequent response 2, contains the remaining data from the

Reviewer: ```suggestion
Subsequent response 2 contains the remaining data from the
```

# E. Appendix: Examples

## Incremental Delivery Examples

### Example 1 - A query containing both defer and stream

```graphql example
query {
  person(id: "cGVvcGxlOjE=") {
    ...HomeWorldFragment @defer(label: "homeWorldDefer")
    name
    films @stream(initialCount: 1, label: "filmsStream") {
      title
    }
  }
}
fragment HomeWorldFragment on Person {
  homeWorld {
    name
  }
}
```

The response to this request will be an _incremental stream_ consisting of an
_initial incremental stream result_ followed by one or more _incremental stream
update result_.

The _initial incremental stream result_ has:

- a {"data"} entry containing the results of the GraphQL operation except for
  the `@defer` and `@stream` selections;
- a {"pending"} entry containing two _pending result_, one for the `@defer`
  selection and for the the `@stream` selection, indicating that these results
  will be delivered in a later _incremental stream update result_;
- a {"hasNext"} entry with the value {true}, indicating that the response is not
  yet complete.

> If an error were to occur, it would also have an {"error"} entry; but not in

Reviewer 2: ```suggestion
If an error were to occur, it would also have an {"errors"} entry; but not in
```

this example.

```json example
{
  "data": {
    "person": {
      "name": "Luke Skywalker",
      "films": [{ "title": "A New Hope" }]
    }
  },
  "pending": [
    { "id": "0", "path": ["person"], "label": "homeWorldDefer" },
    { "id": "1", "path": ["person", "films"], "label": "filmsStream" }
  ],
  "hasNext": true
}
```

Depending on the behavior of the backend and the time at which the deferred and
streamed resources resolve, the stream may produce results in different orders.
In this example, our first _incremental stream update result_ contains the

> deferred data and the first streamed list item. There is one _completed result_,

Reviewer 2: Maybe change "completed result" to "incremental completion result" to tie it in with the "incremental" naming? (Also to be more specific.)

```suggestion
deferred data and the first streamed list item. There is one _incremental completion result_,
```

Author: Updated now to _incremental completion notice_

indicating that the deferred data has been completely delivered.

```json example
{
  "incremental": [
    {
      "id": "0",
      "data": { "homeWorld": { "name": "Tatooine" } }
    },
    {
      "id": "1",
      "items": [{ "title": "The Empire Strikes Back" }]
    }
  ],
  "completed": [
    {"id": "0"}
  ]
  "hasNext": true
}
```

The second _incremental stream update result_ contains the final stream results.
In this example, the underlying iterator does not close synchronously so
{"hasNext"} is set to {true}. If this iterator did close synchronously,
{"hasNext"} would be set to {false} and this would be the final incremental

> stream update result.

Reviewer 2: I think "could" rather than "would" to allow vendor flexibility?

```suggestion
{"hasNext"} is set to {true}. If this iterator did close synchronously,
{"hasNext"} could be set to {false} and make this the final incremental
stream update result.
```

```json example
{
  "incremental": [
    {
      "id": "1",
      "items": [{ "title": "Return of the Jedi" }]
    }
  ],
  "hasNext": true
}
```

The third _incremental stream update result_ contains no incremental data.
{"hasNext"} set to {false} indicates the end of the _incremental stream_. This
incremental stream update result is sent when the underlying iterator of the

> `films` field closes.

Reviewer 2: ```suggestion
When the underlying iterator of the `films` field closes there is no more data
to deliver, so the third and final _incremental stream update result_ sets
{"hasNext"} to {false} to indicate the end of the _incremental stream_.
```

returned in the initial incremental stream result because it is also defined
outside of any fragments with the `@defer` directive. Additionally, there are
two _pending result_ indicating that results for both `@defer`s in the query
will be delivered in later _incremental stream update result_.

```json example
{
  "data": {
    "person": {
      "firstName": "Luke"
    }
  },
  "pending": [
    { "id": "0", "path": ["person"], "label": "homeWorldDefer" },
    { "id": "1", "path": ["person"], "label": "nameAndWorld" }
  ],
  "hasNext": true
}
```

In this example, the first _incremental stream update result_ contains the
deferred data from `HomeWorldFragment`. There is one _completed result_,
indicating that `HomeWorldFragment` has been completely delivered. Because the
`homeWorld` field is present in two separate `@defer`s, it is separated into its
own _incremental result_.

The second _incremental result_ in this _incremental stream update result_
contains the data for the `terrain` field. This _incremental result_ contains a
{"subPath"} entry to indicate to clients that the _response position_ of this
result can be determined by concatenating the path from the _pending result_

> with id `"0"` and the value of this {"subPath"} entry.

Reviewer 2: This reads a little ambiguously: one misinterpretation would be "concatenating (the path from the pending result) with (id "0") and (the value of this subpath entry)".

I've tweaked the grammar slightly to avoid this:

```suggestion
result can be determined by concatenating: the path from the _pending result_
for id `"0"`, and the value of this {"subPath"} entry.
```

```json example
{
  "incremental": [
    {
      "id": "0",
      "data": { "homeWorld": { "name": "Tatooine" } }

>     },

Reviewer 2: I feel like the weirdness of this needs more calling out - name belongs to both id:0 and id:1, so we're kind of using id:0 as a shortcut (but id:1 would also work). We should make it clear to a client that it can't form `HomeWorldFragment` from solely the root data plus the `id:0` stuff and `NameAndHomeWorldFragment` from solely the root data and the `id:1` stuff - they must merge the trees before extracting the data.

Author: Added a paragraph here: https://github.com/graphql/graphql-spec/pull/1203/changes#diff-0237c26b29c484df8d43b13dc3a3658867cfd015d8ccea8ed93454c466dbacb3R200-R203
